tsfc/fem.py

```diff
@@ -323,12 +314,33 @@ def point_expr(self):
     def weight_expr(self):
         return self.quadrature_rule.weight_expression
 
-    @serial_cache(hashkey=make_basis_evaluation_key)
+    @staticmethod
+    def _make_basis_evaluation_key(finat_element, mt, entity_id):
+        ufl_element = mt.terminal.ufl_element()
+        domain = extract_unique_domain(mt.terminal)
+        coordinate_element = domain.ufl_coordinate_element()
+        # This way of caching is fragile.
+        # Should implement _hash_key_() in ModifiedTerminal and include the entire mt in the key,
+        # or only pass necessary bits in mt to basis_evaluation.
+        return (ufl_element, mt.local_derivatives, entity_id, coordinate_element, mt.restriction, domain._ufl_hash_data_())
```

Comment:
Are we using `mt.restriction` anywhere?

Reply:
We can do that in a separate PR and remove those comments, but the fix in this PR is the only thing that I need immediately.